### PR TITLE
Fix display bug when clearing null character + improve sync time.

### PR DESCRIPTION
**Pull request by tmiw**
As part of #97, this fixes a display bug I noticed while testing PSK Reporter functionality (and may also fix the occasional segfault others have noticed too based on the ones I've gotten). I also made some tweaks to improve sync time for the text field when PSK Reporter functionality is turned on.

### CODE REVIEW

**Comment by tmiw** [comment]
@drowe67, should the PDF be automatically be getting generated when the manual isn't updated as part of a PR?

**Comment by drowe67** [comment]
> @drowe67, should the PDF be automatically be getting generated when the manual isn't updated as part of a PR?

No, the [github action](https://github.com/drowe67/freedv-gui/blob/master/.github/workflows/update_freedv_manual.yml#L9) should only update the PDF if `USER_MANUAL.md` changes, I must have messed it up somehow.

**Comment by tmiw** [comment]
@drowe67, actually, I did merge down from master, which seems to have brought the updated manual. That's probably what triggered the re-generation. Still, that in itself probably shouldn't have triggered the generation if the file was just fast-forwarded.

File: src/callsign_encoder.cpp [hunk 1]
```diff
@@ -140,35 +146,67 @@ void CallsignEncoder::pushReceivedByte(char incomingChar)
...
+                {
+                    *pReceivedCallsign_ = 0;
+                    pReceivedCallsign_ = &receivedCallsign_[0];
+                }
             }
-            else if (decodedStr[0] != '\0' && decodedStr[0] != 0x7F)
+            else if (decodedStr[0] != '\0')
             {
-                // Ignore nulls and sync characters.
+                // Ignore incoming nulls but wipe anything to the right of the current pointer
```
Comment (tmiw) [inline src/callsign_encoder.cpp:163]: @drowe67, I think this is what Kanda was running into. Something about what he was receiving would eventually cause the null to be overwritten and the application to overrun the buffer with a later operation (such as strlen). Unfortunately it's hard to reproduce but I'm fine merging this up and including it in the next test build to see if it helps at all.

**Comment by tmiw** [comment]
@drowe67, BTW, I may try rebasing instead of merging for future PRs. It might help with avoiding regeneration of USER_MANUAL.pdf if it's not strictly needed. 👍

**Comment by drowe67** [comment]
> @drowe67, BTW, I may try rebasing instead of merging for future PRs. It might help with avoiding regeneration of USER_MANUAL.pdf if it's not strictly needed. +1

Yes that's an annoying quirk of an other wise cool feature (auto manual building).  We could try a fix in the GitHub action trigger.  I think it [currently fires](https://github.com/drowe67/freedv-gui/blob/master/.github/workflows/update_freedv_manual.yml) when there is a  change in `USER_MANUAL.md`.

**Comment by tmiw** [comment]
FYI, I ended up getting the crash that Kanda saw while testing another branch without this change:

```
Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Application Specific Information:
stack buffer overflow

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff20511462 __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff2053f610 pthread_kill + 263
2   libsystem_c.dylib             	0x00007fff204927ba __abort + 139
3   libsystem_c.dylib             	0x00007fff2046c8eb __stack_chk_fail + 100
4   org.freedv.freedv             	0x00000001002ea0bd CallsignEncoder::isCallsignValid() const + 285
5   org.freedv.freedv             	0x00000001002a80da MainFrame::OnTimer(wxTimerEvent&) + 6010 (main.cpp:1140)
6   org.freedv.freedv             	0x000000010089600c wxEvtHandler::SearchDynamicEventTable(wxEvent&) + 284
7   org.freedv.freedv             	0x0000000100895d3b wxEvtHandler::ProcessEventLocally(wxEvent&) + 59
8   org.freedv.freedv             	0x0000000100895bf0 wxEvtHandler::ProcessEvent(wxEvent&) + 96
9   org.freedv.freedv             	0x000000010089627c wxEvtHandler::SafelyProcessEvent(wxEvent&) + 12
10  org.freedv.freedv             	0x000000010083e8aa wxTimerImpl::SendEvent() + 90
11  com.apple.CoreFoundation      	0x00007fff2065090d __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 20
12  com.apple.CoreFoundation      	0x00007fff206503e8 __CFRunLoopDoTimer + 922
13  com.apple.CoreFoundation      	0x00007fff2064ff42 __CFRunLoopDoTimers + 307
14  com.apple.CoreFoundation      	0x00007fff2063657f __CFRunLoopRun + 2008
15  com.apple.CoreFoundation      	0x00007fff206356ce CFRunLoopRunSpecific + 563
16  com.apple.HIToolbox           	0x00007fff288bd630 RunCurrentEventLoopInMode + 292
17  com.apple.HIToolbox           	0x00007fff288bd42c ReceiveNextEventCommon + 709
18  com.apple.HIToolbox           	0x00007fff288bd14f _BlockUntilNextEventMatchingListInModeWithFilter + 64
19  com.apple.AppKit              	0x00007fff22e559b1 _DPSNextEvent + 883
20  com.apple.AppKit              	0x00007fff22e54177 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1366
21  com.apple.AppKit              	0x00007fff22e4668a -[NSApplication run] + 586
22  org.freedv.freedv             	0x000000010062c5fb wxGUIEventLoop::OSXDoRun() + 171
23  org.freedv.freedv             	0x0000000100869f4d wxCFEventLoop::DoRun() + 29
24  org.freedv.freedv             	0x00000001007c50fe wxEventLoopBase::Run() + 158
25  org.freedv.freedv             	0x0000000100790f03 wxAppConsoleBase::MainLoop() + 99
26  org.freedv.freedv             	0x00000001005e05ea wxApp::OnRun() + 26
27  org.freedv.freedv             	0x0000000100804878 wxEntry(int&, wchar_t**) + 56
28  org.freedv.freedv             	0x000000010029df73 main + 35 (main.cpp:178)
29  libdyld.dylib                 	0x00007fff2055a621 start + 1
```

Based on the above stack trace, it definitely looks like the prior implementation of isCallsignValid() was causing issues (likely by strcpy() copying way more than it should have).

**Comment by drowe67** [comment]
Note to self - be good to write unit tests for chunks of freedv-gui functionality to test the outside the GUI, for example `freedv_interface.cpp` and PSK reporter.  I'm also wondering if we can drive freedv-gui with sound from virtual sound cards as part of an automated test.

**Comment by tmiw** [comment]
@drowe67, re: virtual sound cards, I think such a UT type thing might be pretty straightforward using PulseAudio. However, it should theoretically work with PortAudio as is since many users already use virtual sound cards for e.g. SDRs.

**Comment by tmiw** [comment]
I started back on this and honestly, I'm not 100% sure I want to keep the code in here. I'll create a new branch/PR in drowe67/freedv-gui and include a different implementation.

**Comment by drowe67** [comment]
@tmiw - no problem, should we close this PR?

**Comment by tmiw** [comment]
@drowe67, yep, go ahead and close.

**Pull request closed without merging**